File: src/main/java/com/murex/tbw/domain/book/Author.java
/*******************************************************************************
 *
 * Copyright (c) {2003-2019} Murex S.A.S. and its affiliates.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 *******************************************************************************/

package com.murex.tbw.domain.book;

import com.murex.tbw.domain.country.Country;

import java.util.Objects;

public final class Author {

    private final String name;
    private final Country nationality;

    public Author(String name, Country nationality) {
        this.name = name;
        this.nationality = nationality;
    }

    public String getName() {
        return name;
    }

    public Country getNationality() {
        return nationality;
    }

    @Override
    public String toString() {
        return "Author{" +
                "name='" + name + '\'' +
                ", nationality=" + nationality +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Author author = (Author) o;
        return Objects.equals(name, author.name) &&
                Objects.equals(nationality, author.nationality);
    }

    @Override
    public int hashCode() {
        return Objects.hash(name, nationality);
    }
}

File: csharp/Application/Domain/Book/Author.cs
/*******************************************************************************
 *
 * Copyright (c) {2003-2019} Murex S.A.S. and its affiliates.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the MIT License
 * which accompanies this distribution, and is available at
 * https://opensource.org/licenses/MIT
 *
 *******************************************************************************/

using System;

namespace Application.Domain.Book
{
    public sealed class Author
    {
        public string Name { get; }
        public Country.Country Nationality { get; }

        public Author(string name, Country.Country nationality)
        {
            Name = name;
            Nationality = nationality;
        }

        public override string ToString() => $"Author [ {nameof(Name)}: '{Name}', " +
                                             $"{nameof(Nationality)}: '{Nationality}' ]";

        private bool Equals(Author other) => Name == other.Name
                                             && Equals(Nationality, other.Nationality);

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != this.GetType()) return false;
            return Equals((Author) obj);
        }

        public override int GetHashCode() => HashCode.Combine(Name, Nationality);
    }
}
